feat: update documentation structure, add troubleshooting section (#746)
ProjectLibertyLabs/gateway: docs/src/Troubleshooting.md

# Troubleshooting

So you set up your app or the [social-app-template](https://github.com/ProjectLibertyLabs/social-app-template), and your posts are not showing up. What next?

Here is a basic walk-through on figuring out where things went wrong.

We will examine the logs from the services configured for the default setup running against the Paseo testnet. We will look at the flow of a post as it starts from the front end, gets published on-chain, and is displayed in a feed on the front end.

Some of these troubleshooting steps assume that the social-app-template is running your front end. If you have implemented your own front end, the Gateway service message flow will be similar.

## Registering Users

### Registration Process

> Provider registration is required to setup a Provider on Frequency. This is a 1-time process Providers use to register, and get a Provider ID.  On Testnet, anyone may set up a Provider.  However, if you wish to set up a Provider on Mainnet, this has to be done in concert with the Frequency Foundation. See the [Provider Dashboard](https://provider.frequency.xyz) for more information.

Reviewer: ```suggestion
Provider registration is required to setup a Provider on Frequency. This is a one-time process Providers use to register and obtain a Provider ID.  Registration of Providers on Testnet is unrestricted; on Mainnet it requires approval from the Frequency Network Foundation. See the [Provider Dashboard](https://provider.frequency.xyz) for more information.
```

> The Register User flow is the process by which a user registers with a Provider.  This is a 1-time process for each user.  The user will be able to log in to the Provider's app using the Sign In With Frequency [SIWF](https://github.com/ProjectLibertyLabs/siwf) process.  The user will now be able to interact with the Provider's app.

Reviewer: ```suggestion
The Register User flow is the process by which a user registers with a Provider.  This is a one-time process required each time a user signs up with a new Provider.  The user will be able to log in to the Provider's app using the Sign In With Frequency [SIWF](https://github.com/ProjectLibertyLabs/siwf) process.  The user will now be able to interact with the Provider's app.
```

```mermaid
sequenceDiagram
    participant UI as User Interface
    participant API as API Service
    participant WS as Worker Service
    participant BL as Blockchain Service

    %% Register User
    Note over UI, API: Register User
    UI->>API: Register User
    API-->>+WS: Register User
    WS-->>BL: Create User With Delegations to Provider
    BL-->>WS: User Created Transaction Finalized
    WS-->>-UI: User Created

    %% Login User
    Note over UI, API: Login User
    UI->>API: Login User
    API-->>WS: Login User
    WS-->>BL: Get User Info
    BL-->>WS: User Info
    WS-->>UI: User Info
```

In debugging the user registration process using SIWF, we are mostly concerned with the Register User section of this diagram, and the flow of events through the Account Service.
The Account Service is composed of two main components: the API Service and the Worker Service.
The API Service handles incoming HTTP requests, while the Worker Service processes jobs that require blockchain interaction.

Here are the steps:

> 1. Provider’s User Interface will interact with Frequency Access to collect the user's identity information and use that information to send a registration request to the Gateway-Account Service API.

Reviewer: ```suggestion
1. Provider’s application UI will invoke the Frequency Access UI, which collects the user's identity information and returns a signed registration payload. The Provider application then uses that information to send a registration request to the Gateway-Account Service API.
```

> 2. Account Service API receives the [SIWF POST](https://projectlibertylabs.github.io/gateway/account/#tag/v2accounts/operation/AccountsControllerV2_postSignInWithFrequency) at `/v2/accounts/swif` and sends them to the Account Service Worker.
>     - Account Service API will respond with a `200 OK` if the request is successful.

Reviewer: ```suggestion
2. Account Service API receives the [SIWF POST](https://projectlibertylabs.github.io/gateway/account/#tag/v2accounts/operation/AccountsControllerV2_postSignInWithFrequency) at `/v2/accounts/swif` and enqueues a request to the Account Service Worker.
    - Account Service API will respond with a `200 OK` if enqueueing the request is successful.
```

3. Account Service Worker has been configured by the Provider to use Capacity to process the transactions and send them to the blockchain.
    - Account Service Worker will process the request and send a transaction to the blockchain.

>     - Accound Service Worker will montior the blockchain for the transaction to be finalized.

Reviewer: ```suggestion
    - Accound Service Worker will monitor the blockchain for the transaction to be finalized.
```

> 4. The blockchain finalizes the transactions and Account Service Worker detects the finalized transaction. The Provider application may choose from several different methods to received notification that the transaction has been finalized:

Reviewer: ```suggestion
4. The blockchain finalizes the transactions and Account Service Worker detects the finalized transaction. The Provider application may choose from several different methods to receive notification that the transaction has been finalized:
```

#         The following logs show the Account Service Worker sending a webhook callback to the Provider,
#         this should match your webhook URL that you configured for the Gateway Account Service.]
# [...]
  DEBUG [TxnNotifierService] Sending transaction notification to webhook: http://mock-webhook-logger:3001/webhooks/account-service
  DEBUG [TxnNotifierService] Transaction: {"transactionType":"SIWF_SIGNUP","providerId":"729","referenceId":"7lsjlxLT2ImiRLkwk8vaE-aWq8U","msaId":"13232","accountId":"5GBa7oCJFkTWMNLU8w43FSzhv66kR9kpHD5BecnG3Dyx2Dny","handle":"gateway-16.49"}
```

#### When Things Go Wrong

Look in these logs for any errors or warnings. If you see any, they will give you a clue as to what went wrong.

Use the block number to look up the transaction on the [Polkadot blockchain explorer](https://polkadot.js.org/apps/?rpc=wss%3A%2F%2F0.rpc.testnet.amplica.io#/explorer).
This will give you more information about the transaction and the error that occurred.

<!-- {{#png-embed ./src/PolkadotBlockExplorer.png Polkadot Block Explorer}} -->
![Polkadot Block Explorer](./PolkadotBlockExplorer.png)

In the above image you will able to see the transactions in that block.

> **NOTE:** If you are on testnet, you may not be the only contributor of transactions to that block.
> Anybody can be posting to that block.  However, you should be able to find the transaction within that block by

Reviewer: ```suggestion
**NOTE:** If you are on a public chain (ie, testnet or mainnet), you may not be the only contributor of transactions to that block.
Anybody can be posting to that block.  However, you should be able to find the transaction within that block by
```

 1) Mapping to the correct extrinsics (msa.msaCreated, msa.DelegationGranted, msa.HandleClaimed and statefulStorage.ItemizedPageUpdated)
 2) Finding the transaction that maps to your provider MSA (729 in the example above)
 3) Checking the User Key (5GBa7oCJF… in the example below)
 4) Checking the user handle provided to you against the one claimed in the transaction (gateway-16.49 in the example below)

You may get a message that the extrinsic failed.
In this case, there will be an error code you can reference.

> In the below example, the error code is 60.

Reviewer: ```suggestion
In the below example, the error index is 60 and the error code is 0. Error index indicates the Frequency Substrate pallet that produced the error, and the error code is the value of the error enum within that pallet. (Pallet indices can be found [here](https://github.com/frequency-chain/frequency/blob/00c3a2fdc97a4a7448577faf4af445fb08e8c1b0/runtime/frequency/src/lib.rs#L1285). Error codes for each pallet are typically found within the pallet's main `lib.rs` source (e.g., the `msa` pallet [error code 0](https://github.com/frequency-chain/frequency/blob/00c3a2fdc97a4a7448577faf4af445fb08e8c1b0/pallets/msa/src/lib.rs#L297) indicates `KeyAlreadyRegistered`
```

Reviewer 2: ```suggestion
In the below example, the error index is 60 and the error code is 0. Error index indicates the Frequency Substrate pallet that produced the error, and the error code is the value of the error enum within that pallet. A list of errors can be found in [the Frequency Wiki](https://github.com/frequency-chain/frequency/wiki/Errors).
```
